core/src/main/java/org/apache/iceberg/PartitionsTable.java

```diff
@@ -127,17 +127,20 @@ private DataTask task(StaticTableScan scan) {
   }
 
   /**
-   * This method transforms the table's partition spec to a spec that is used to evaluate
-   * the user-provided filter expression against the 'partitions' table.
+   * This method transforms the table's partition spec to a spec that is used to rewrite the user-provided filter
+   * expression against the partitions table.
+   * <p>
+   * The resulting partition spec maps partition.X fields to partition X using an identity partition transform. When
+   * this spec is used to project an expression for the partitions table, the projection will remove predicates for
+   * non-partition fields (not in the spec) and will remove the "partition." prefix from fields.
    *
-   * In particular, it adds 'partition.' prefix to each of the table's partition field
-   * (ie, x => partition.x)
-   * @param spec table spec
-   * @return a spec that can be used on the partitions table.
+   * @param partitionTableSchema schema of the partition table
+   * @param spec spec on which the partition table schema is based
+   * @return a spec used to rewrite partition table filters to partition filters using an inclusive projection
    */
-  private PartitionSpec transformSpec(PartitionSpec spec) {
-    PartitionSpec.Builder identitySpecBuilder = PartitionSpec.builderFor(schema());
-    spec.fields().forEach(pf -> identitySpecBuilder.identity(PARTITION_FIELD_PREFIX + pf.name()));
+  private static PartitionSpec transformSpec(Schema partitionTableSchema, PartitionSpec spec) {
+    PartitionSpec.Builder identitySpecBuilder = PartitionSpec.builderFor(partitionTableSchema);
+    spec.fields().forEach(pf -> identitySpecBuilder.identity(PARTITION_FIELD_PREFIX + pf.name(), pf.name()));
```

Comment:
The key here is that the schema's field name is `partition.X` but the name of the partition is just `X`. That way, the projection will automatically remove the `partition.` prefix and will drop any expressions for non-partition fields.